### PR TITLE
Fix exception checks in the tests

### PR DESCRIPTION
**Pull request by psfinaki**
Found when reviewing [this PR](https://github.com/dotnet/fsharp/pull/17277). The code throws `ArgumentOutOfRangeException` and the tests verify `ArgumentException`, [here](https://github.com/dotnet/fsharp/pull/17277/files#diff-c73df57d61503610bcde77c681a77284c316ea919357288192e29152abe5a0f5R1323). Not good.

`CheckThrowsExn` should be adjusted to match the exception type, otherwise it only checks them within the type hierarchy.

### CODE REVIEW

File: tests/FSharp.Core.UnitTests/LibraryTestFx.fs [hunk 1]
```diff
@@ -18,7 +18,7 @@ let CheckThrowsExn<'a when 'a :> exn> (f : unit -> unit) =
         let _ = f ()
         sprintf "Expected %O exception, got no exception" typeof<'a> |> Assert.Fail 
     with
-    | :? 'a -> ()
+    | :? 'a as e when e.GetType() = typeof<'a> -> ()
```
Comment (T-Gro) [inline tests/FSharp.Core.UnitTests/LibraryTestFx.fs:21]: TBH I do not think that it is wrong when an API promises is to throw exception A, and instead it returns B which inherits A.
This should be an implementation detail, and as such should probably not cause tests to fail.

Comment (T-Gro) [reply tests/FSharp.Core.UnitTests/LibraryTestFx.fs:21]: I assume this is what you used when figuring out which tests to change in the first place.
But now that it has happened, I think the whole when guard can be removed.

Comment (T-Gro) [inline tests/FSharp.Core.UnitTests/LibraryTestFx.fs:21]:
```suggestion
    | :? 'a as e  -> ()
```

File: tests/FSharp.Core.UnitTests/FSharp.Core/OperatorsModuleDynamic.fs [hunk 2]
```diff
@@ -18,7 +18,7 @@ module OperatorsModuleDynamic =
             let _ = f ()
             sprintf "Expected %O exception, got no exception" typeof<'a> |> Assert.Fail 
         with
-        | :? 'a -> ()
+        | :? 'a as e when e.GetType() = typeof<'a> -> ()
```
Comment (T-Gro) [inline tests/FSharp.Core.UnitTests/FSharp.Core/OperatorsModuleDynamic.fs:21]:
```suggestion
        | :? 'a -> ()
```

**Review by KevinRansom — APPROVED** [review]

File: tests/FSharp.Core.UnitTests/LibraryTestFx.fs (hunk 1 above)
Comment (psfinaki) [reply tests/FSharp.Core.UnitTests/LibraryTestFx.fs:21]: That was my first thought also. But now I believe it rather applies to the production code (and hence the operator :? was likely designed to follow the inheritance chain), whereas in these test frameworks we should be more precise.

Getting back to the original trigger (the PR I was reviewing), we have the function taking an argument and we want to:
1) Throw ArgumentNullException for value X
2) Throw ArgumentOutOfBoundsException for value Y
3) Throw ArgumentException for value Z

I want my tests to be expressive about that with code like
```fsharp
CheckThrowsArgumentNullException (fun () -> (fun () -> func(X))
CheckThrowsArgumentOutOfBoundsException (fun () -> func(Y))
CheckThrowsArgumentException (fun () -> func(Z))
```
If this above does NOT actually check the exact exception, how to ensure the precise behavior? We'd either need to review code manually or force end tests to add a type check to every case like this.

To sum up, I think for the case of test frameworks, this precise check is beneficial.

Comment (T-Gro) [reply tests/FSharp.Core.UnitTests/LibraryTestFx.fs:21]: It can check the exception and allow either that OR allow any of its inherited types, because LSP should still hold here.

If it checks CheckThrowsArgumentOutOfBoundsException and something which inherits ArgumentOutOfBoundsException  -> that's OK.

Comment (T-Gro) [reply tests/FSharp.Core.UnitTests/LibraryTestFx.fs:21]: If this is kept and lib code is reimplemented, while API still holds - this is just sunk costs to update the tests, even when there is nothing wrong with such code.
The test should only fail if something goes wrong in the impl, not when it changes internal implementation detail.

Comment (psfinaki) [reply tests/FSharp.Core.UnitTests/LibraryTestFx.fs:21]: Took me a while to understand why you're mentioning LSP :D

No I understand what you mean, it's more important that we detect when `ArgumentException` is thrown instead of `ArgumentOutOfBoundsException` and not the opposite case - and the current (unmodified) impl does check this.

I just don't see anything bad in being more precise here as it will help avoid similar situations in the future and will force tests to read closer to the spec.

Comment (T-Gro) [reply tests/FSharp.Core.UnitTests/LibraryTestFx.fs:21]: Because there is nothing "precise" about it at all.
If API says it returns `A`, and instead it returns `B : A` - that is still very precise meeting of the spec, and this is the case across many platforms incl. dotnet.

The only thing insisting on exact type and disallowing subtypes can cause is additional effort for people who would implement an actual good change and the tests would get in the way.

And yes, it is this LSP which was meant by it: https://en.wikipedia.org/wiki/Liskov_substitution_principle :))

**Review by abonie — APPROVED** [review]

File: tests/FSharp.Core.UnitTests/LibraryTestFx.fs (hunk 1 above)
Comment (psfinaki) [reply tests/FSharp.Core.UnitTests/LibraryTestFx.fs]: So - not convinced, but 2:1 and the point was to fix the tests, so let's keep things as they are :)

**Review by T-Gro — APPROVED** [review]
